src/ipips/ipip-0499.md

```diff
@@ -0,0 +1,132 @@
+relatedIssues:
+  - https://discuss.ipfs.tech/t/should-we-profile-cids/18507
+order: 0499
+tags: ['ipips']
+---
+
+## Summary
+
+This proposal introduces configuration profiles for CIDs used to represent files and directories with UnixFS. These ensure that the deterministic CID generation for the same data, regardless of the implementation.
+
+Profiles explicitly define the UnixFS parameters, e.g. dag width, hash algorithm, and chunk size, that affect the resulting CID, such that given the profile and input data different implementations will generate identical CIDs.
+
+## Motivation
+
+UnixFS CIDs are not deterministic. This means that the same file tree can yield different CIDs depending on the parameters used by the implementation to generate it, which in some cases, aren't even configurable by the user. For example, the chunk size, DAG width, and layout can vary between implementations or even between different versions of the same implementation.
+
+This lack of determinism makes has a number of drawbacks:
+
+- It is difficult to verify content across different tools and implementations, as the same content may yield different CIDs.
+- Users are required to store and transfer UnixFS merkle proofs in order to verify CIDs, adding storage overhead, network bandwidth, and complexity to the verification process.
+- In terms of developer experience, it deviates from the mental model of a hash function, where the same input should always yield the same output. This leads to confusion and frustration when working with UnixFS CIDs
+
+By introducing profiles which define the parameters that affect the root CID of the DAG, we can benefit from both the optionality offered by UnixFS, where users are free to chose their own parameters, and determinism through profiles.
+
+## Detailed design
+
+We introduce a set of named profiles that define a set of parameters for generating UnixFS CIDs. These profiles can be used by implementations to ensure that the same content will yield the same CID across different tools and implementations.
+
+### UnixFS parameters
+
+The profiles define a set of parameters that affect the resulting string encoding of the CID. These parameters are based on the UnixFS specification and are used to generate the CID for a given file tree. The parameters include:
+
+1. CID version, e.g. CIDv0 or CIDv1
+1. Multibase encoding for the CID, e.g. base32
+1. Hash function used for all nodes in the DAG, e.g. sha2-256
+1. UnixFS file chunking algorithm
+1. UnixFS file chunk size or target (if required by the chunking algorithm)
+1. UnixFS DAG layout (e.g. balanced, trickle etc...)
+1. UnixFS DAG width (max number of links per `File` node)
+1. `HAMTDirectory` bitwidth, i.e. the number of bits determines the fanout of the `HAMTDirectory` (default bitwidth is 8 == 256 leaves).
+1. `HAMTDirectory` threshold (max `Directory` size before switching to `HAMTDirectory`): based on an estimate of the block size by counting the size of PNNode.Links
```

Comment:
```suggestion
1. `HAMTDirectory` threshold (max `Directory` size before switching to `HAMTDirectory`): based on an estimate of the block size by counting the size of PNNode.Links. We do not include details about the estimation algorithm as we do not encourage implementations to support it.
```

Reply:
I don't think we should be estimating the block size as it's trivial to calculate it exactly.  Can we not just define this (and punt to the spec for the details) to make it less hand-wavey?

```suggestion
1. `HAMTDirectory` threshold (max `Directory` size before switching to `HAMTDirectory`): based on the final size of the serialized form of the [PBNode protobuf message](https://specs.ipfs.tech/unixfs/#dag-pb-node) that represents the directory.
```